src/target/nrf91.c

```diff
@@ -81,25 +134,180 @@ static void nrf91_add_flash(target_s *target, uint32_t addr, size_t length, size
 	flash->write = nrf91_flash_write;
 	flash->erased = 0xff;
 	target_add_flash(target, flash);
+
+	/* add separate UICR flash */
+	target_flash_s *flash_uicr = calloc(1, sizeof(*flash_uicr));
+	if (!flash_uicr) { /* calloc failed: heap exhaustion */
+		DEBUG_WARN("calloc: failed in %s\n", __func__);
+		return;
+	}
+
+	flash_uicr->start = 0xff8000U;
+	flash_uicr->length = 0x1000U;
+	flash_uicr->blocksize = 0x1000U;
+	flash_uicr->erase = nrf91_uicr_flash_erase;
+	flash_uicr->write = nrf91_flash_write;
+	flash_uicr->erased = 0xff;
+	target_add_flash(target, flash_uicr);
+
+}
+
+static bool nrf91_mass_erase(target_s *target)
+{
+	adiv5_access_port_s *ap = cortex_ap(target);
+	adiv5_access_port_s ctrl_ap = {
+		.dp = ap->dp,
+		.apsel = 0x4U,
+	};
+
+	return nrf91_ctrl_ap_mass_erase(&ctrl_ap);
 }
 
 bool nrf91_probe(target_s *target)
 {
 	adiv5_access_port_s *ap = cortex_ap(target);
 
-	if (ap->dp->version < 2U)
+	if (ap->dp->version < 2U || ap->dp->target_partno != 0x90U)
 		return false;
 
-	switch (ap->dp->target_partno) {
-	case 0x90:
-		target->driver = "Nordic nRF9160";
-		target->target_options |= TOPT_INHIBIT_NRST;
-		target_add_ram(target, 0x20000000, 256U * 1024U);
-		nrf91_add_flash(target, 0, 4096U * 256U, 4096U);
-		break;
-	default:
+	uint32_t buf[3] = {0};
+
+	ap->dp->mem_read(ap, buf, 0x00FF0140, 3*sizeof(uint32_t));
+
+	gdb_outf("nRF%04x %4s%4s detected!\n", buf[0], (const char *)&buf[2], (const char *)&buf[1]);
```

Comment:
This would be fine as a DEBUG_* statement, but should be avoided in the scan output as it's not guaranteed where exactly this will end up in the information display, while also duplicating the information to the user from the target->driver strings.

Same comment as below on uint32_t needing PRIx32.